src/backend/optimizer/prep/prepunion.c

```diff
@@ -1808,6 +1808,54 @@ adjust_appendrel_attrs(PlannerInfo *root, Node *node, AppendRelInfo *appinfo)
 	return result;
 }
 
+typedef struct nested_subplans_context
+{
+	plan_tree_base_prefix base;
+	PlannerInfo *root;
+} nested_subplans_context;
+
+static bool
+nested_subplans(Node *node, adjust_appendrel_attrs_context *context) {
+	nested_subplans_context *new_context = (nested_subplans_context*)malloc(sizeof(nested_subplans_context));
+	new_context->root = context->root;
+
+	nested_subplan_mutator(node, new_context);
+}
+
+static bool
+nested_subplans_mutator(Node *node, adjust_appendrel_attrs_context *context) {
+	if (node == NULL) {
+		return false;
+	}
+
+	if (IsA(node, SubPlan)) {
+		SubPlan *sp = (SubPlan*) node;
+
+		if (!sp->is_initplan) {
```

Comment:
Originally, the planner didn't recurse into initplans because it didn't want to create any copies of subplans inside initplans, because initplans are executed independently. When we copied this logic, we didn't want this either, so the behavior when we avoid initplans is intentional.

Reply:
It seems you are right and gpdb handle possible partition table scan inside initplans independently. So we don't need to recurse into them.